### PR TITLE
fix(gatsby-plugin-image): Correct IntersectionObserver handling

### PR DESCRIPTION
**Pull request by ascorbic**
It turns out that the intersectionobserver never worked proeprly, but our tests didn't catch it because it works for the first image but not subsequent ones, and aside from Safari the images were all loading immediately when they were inside the intersection threshold. The problem was that we were using a shared IO instance (which is fine), but the IO's event listener was only calling the callback of the element that originally created it, and not any that were subsequently observed. This fixes it by using a WeakMap to map the image DOM element to its callback.

This also fixes the intersection thresholds, which were an arbitrary 150%. They now match [the values used by Chrome native lazy loading](https://web.dev/browser-level-image-lazy-loading/#distance-from-viewport-thresholds), which vary according to the connection speed (where available).

### CODE REVIEW

File: packages/gatsby-plugin-image/src/components/intersection-observer.ts [hunk 1]
```diff
@@ -1,12 +1,27 @@
...
 
+const ioEntryMap = new WeakMap<HTMLElement, () => void>()
+/* eslint-disable @typescript-eslint/no-explicit-any  */
+const connection =
+  (navigator as any).connection ||
+  (navigator as any).mozConnection ||
+  (navigator as any).webkitConnection
+/* eslint-enable @typescript-eslint/no-explicit-any */
+
+// These match the thresholds used in Chrome's native lazy loading
```
Comment (wardpeet) [inline packages/gatsby-plugin-image/src/components/intersection-observer.ts:16]: Please add a link to the resource
```suggestion
// These match the thresholds used in Chrome's native lazy loading
// @see https://web.dev/browser-level-image-lazy-loading/#distance-from-viewport-thresholds
```

File: packages/gatsby-plugin-image/src/components/intersection-observer.ts [hunk 2]
```diff
@@ -20,13 +35,18 @@ export function createIntersectionObserver(
...
+            // We only need to call it once
+            ioEntryMap.delete(entry.target as HTMLElement)
           }
         })
       },
       {
-        // TODO look at lighthouse & amp-image on what values they are using
-        rootMargin: `150%`,
+        rootMargin:
+          connectionType === `4g`
```
Comment (wardpeet) [inline packages/gatsby-plugin-image/src/components/intersection-observer.ts:47]: could you also check `connection.saveData`
```suggestion
          connectionType === `4g` && !connection.saveData
```

**Review by wardpeet — APPROVED** [review]
Nice! Could you add an e2e test so we don't have any regressions in the future?

**Comment by mfrachet** [comment]
@wardpeet this is exactly why actually have our conversation about devtools and Cypress 😊 : I'm writing them

**Comment by wardpeet** [comment]
E2e test can be in a follow up, no need to block this PR for it

**Comment by ascorbic** [comment]
Marvin's a machine, and they're already in. I've just asked him to move them, but otherwise they look great to me

File: e2e-tests/gatsby-static-image/cypress/integration/intersection-observer.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+describe(`GatsbyImage / Intersection Observer`, () => {
```
Comment (wardpeet) [inline e2e-tests/gatsby-static-image/cypress/integration/intersection-observer.js:1]:
```suggestion
describe(`GatsbyImage / Below viewport native lazy loading`, () => {
```

**Review by wardpeet — DISMISSED** [review]

**Comment by mfrachet** [comment]
I think we are still testing the intersection observer here. I'm always hitting https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-image/src/components/gatsby-image.browser.tsx#L111 because `hasSSR`is not defined in https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-image/src/components/gatsby-image.browser.tsx#L94

**Comment by ascorbic** [comment]
@mfrachet It'll hit that if you're running gatsby develop, but these tests run gatsby build/serve, so they will have hasSSR set

**Pull request merged**